AssertionError: kind(sig) == K"call"

This seems similar to, but not the same as, https://github.com/JuliaDebug/Cthulhu.jl/issues/399.

This can be reproduced as follows:
```julia
julia> struct TupleWrapper # concrete
           tuple::Tuple    # abstract
       end

julia> using Dates

julia> const JuliaToRelPeriodTypes = Dict{Type{<:Dates.Period}, TupleWrapper}(
           Dates.Year => TupleWrapper((1,2,3)),
           Dates.Month => TupleWrapper((UInt128(1), UInt128(2)))
       )
Dict{Type{<:Period}, TupleWrapper} with 2 entries:
  Year  => TupleWrapper((1, 2, 3))
  Month => TupleWrapper((0x00000000000000000000000000000001, 0x00000000000000000000000000000002))

julia> const RelToJuliaPeriodTypes = Dict(v => k for (k,v) in JuliaToRelPeriodTypes)
Dict{TupleWrapper, DataType} with 2 entries:
  TupleWrapper((0x00000000000000000000000000000001, 0x00000000000000000000000000000002)) => Month
  TupleWrapper((1, 2, 3))                                                                => Year

julia> foo(x) = JuliaToRelPeriodTypes[x]
foo (generic function with 1 method)

julia> using Cthulhu

julia> @descend foo(Dates.Year)
foo(x) in Main at REPL[5]:1
1 foo(x::Type{Year})::TupleWrapper = JuliaToRelPeriodTypes::Dict{Type{<:Period}, TupleWrapper}[x::Type{Year}]
Select a call to descend into or ↩ to ascend. [q]uit. [b]ookmark.
Toggles: [w]arn, [h]ide type-stable statements, [t]ype annotations, [s]yntax highlight for Source/LLVM/Native.
Show: [S]ource code, [A]ST, [T]yped code, [L]LVM IR, [N]ative code
Actions: [E]dit source code, [R]evise and redisplay
 • 1 JuliaToRelPeriodTypes::Dict{Type{<:Period}, TupleWrapper}[x::Type{Year}]
   ↩
[ Info: tracking Base
ERROR: AssertionError: kind(sig) == K"call"
Stacktrace:
  [1] num_positional_args(tsn::JuliaSyntax.TreeNode{TypedSyntax.TypedSyntaxData})
    @ TypedSyntax ~/.julia/packages/TypedSyntax/VbsHx/src/node.jl:339
  [2] truncate_if_defaultargs!(tsn::JuliaSyntax.TreeNode{TypedSyntax.TypedSyntaxData}, mappings::Vector{Vector{Union{JuliaSyntax.TreeNode{JuliaSyntax.SyntaxData}, JuliaSyntax.TreeNode{TypedSyntax.TypedSyntaxData}}}}, meth::Method)
    @ Cthulhu ~/.julia/packages/Cthulhu/tlU2N/src/reflection.jl:367
  [3] get_typed_sourcetext(mi::Core.MethodInstance, src::Core.CodeInfo, rt::Any; warn::Bool)
    @ Cthulhu ~/.julia/packages/Cthulhu/tlU2N/src/reflection.jl:355
  [4] find_callsites(interp::Cthulhu.CthulhuInterpreter, CI::Core.CodeInfo, stmt_infos::Vector{Any}, mi::Core.MethodInstance, slottypes::Vector{Any}, optimize::Bool, annotate_source::Bool)
    @ Cthulhu ~/.julia/packages/Cthulhu/tlU2N/src/reflection.jl:28
  [5] _descend(term::REPL.Terminals.TTYTerminal, interp::Cthulhu.CthulhuInterpreter, curs::Cthulhu.CthulhuCursor; override::Nothing, debuginfo::Cthulhu.DInfo.DebugInfo, optimize::Bool, interruptexc::Bool, iswarn::Bool, hide_type_stable::Bool, verbose::Nothing, remarks::Bool, with_effects::Bool, inline_cost::Bool, type_annotations::Bool, annotate_source::Bool)
    @ Cthulhu ~/.julia/packages/Cthulhu/tlU2N/src/Cthulhu.jl:497
  [6] _descend(term::REPL.Terminals.TTYTerminal, interp::Cthulhu.CthulhuInterpreter, curs::Cthulhu.CthulhuCursor; override::Nothing, debuginfo::Symbol, optimize::Bool, interruptexc::Bool, iswarn::Bool, hide_type_stable::Bool, verbose::Nothing, remarks::Bool, with_effects::Bool, inline_cost::Bool, type_annotations::Bool, annotate_source::Bool)
    @ Cthulhu ~/.julia/packages/Cthulhu/tlU2N/src/Cthulhu.jl:630
  [7] _descend(term::REPL.Terminals.TTYTerminal, interp::Cthulhu.CthulhuInterpreter, mi::Core.MethodInstance; kwargs::Base.Pairs{Symbol, Bool, Tuple{Symbol}, NamedTuple{(:iswarn,), Tuple{Bool}}})
    @ Cthulhu ~/.julia/packages/Cthulhu/tlU2N/src/Cthulhu.jl:766
  [8] _descend(term::REPL.Terminals.TTYTerminal, args::Any; interp::Core.Compiler.NativeInterpreter, kwargs::Base.Pairs{Symbol, Bool, Tuple{Symbol}, NamedTuple{(:iswarn,), Tuple{Bool}}})
    @ Cthulhu ~/.julia/packages/Cthulhu/tlU2N/src/Cthulhu.jl:782
  [9] __descend_with_error_handling(args::Any; terminal::Any, kwargs::Base.Pairs{Symbol, V, Tuple{Vararg{Symbol, N}}, NamedTuple{names, T}} where {V, N, names, T<:Tuple{Vararg{Any, N}}})
    @ Cthulhu ~/.julia/packages/Cthulhu/tlU2N/src/Cthulhu.jl:240
 [10] _descend_with_error_handling(f::Any, argtypes::Any; kwargs::Base.Pairs{Symbol, Bool, Tuple{Symbol}, NamedTuple{(:iswarn,), Tuple{Bool}}})
    @ Cthulhu ~/.julia/packages/Cthulhu/tlU2N/src/Cthulhu.jl:229
 [11] descend_code_typed(::Any, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Cthulhu ~/.julia/packages/Cthulhu/tlU2N/src/Cthulhu.jl:174
 [12] descend_code_typed(::Any, ::Any)
    @ Cthulhu ~/.julia/packages/Cthulhu/tlU2N/src/Cthulhu.jl:174
 [13] top-level scope
    @ REPL[7]:1

julia>
```

Here are my version infos:
```julia
julia> versioninfo()
Julia Version 1.8.2
Commit 36034abf26 (2022-09-29 15:21 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin22.3.0)
  CPU: 12 × Apple M2 Max
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, apple-m1)
  Threads: 1 on 8 virtual cores
```
```julia
(@v1.8) pkg> st -m
Status `~/.julia/environments/v1.8/Manifest.toml`
  [1520ce14] AbstractTrees v0.4.4
  [c7e460c6] ArgParse v1.1.4
  [69666777] Arrow v2.5.0
  [31f734f8] ArrowTypes v2.0.2
  [6e4b80f9] BenchmarkTools v1.3.2
  [d1d4a3ce] BitFlags v0.1.7
  [c3b6d118] BitIntegers v0.2.7
  [e1450e63] BufferedStreams v1.1.0
  [fa961155] CEnum v0.4.2
  [85eb1798] CloudBase v1.3.5
  [3365d9ee] CloudStore v1.3.9 `~/.julia/dev/CloudStore`
  [da1fd8a2] CodeTracking v1.3.0
  [5ba52731] CodecLz4 v0.4.0
  [944b1d66] CodecZlib v0.7.1
  [6b39b394] CodecZstd v0.7.2
  [3da002f7] ColorTypes v0.11.4
  [5ae59095] Colors v0.12.10
  [34da2185] Compat v4.6.1
  [88353bc9] ConfParser v0.1.2
  [f68482b8] Cthulhu v2.8.11
  [9a962f9c] DataAPI v1.14.0
  [e2d170a0] DataValueInterfaces v1.0.0
  [4e289a0a] EnumX v1.0.4
  [460bff9d] ExceptionUnwrapping v0.1.6
  [e2ba6199] ExprTools v0.1.9
  [8f5d6c58] EzXML v1.1.0
  [937e9308] Figgy v1.1.0
  [5789e2e9] FileIO v1.16.0
  [53c48c17] FixedPointNumbers v0.8.4
  [08572546] FlameGraphs v1.0.0
  [1eca21be] FoldingTrees v1.2.1
  [069b7b12] FunctionWrappers v1.1.3
  [cd3eb016] HTTP v1.7.4
  [9b13fd28] IndirectArrays v1.0.0
  [83e8ac13] IniFile v0.5.1
  [842dd82b] InlineStrings v1.4.0
  [c8e1da08] IterTools v1.4.0
  [82899510] IteratorInterfaceExtensions v1.0.0
  [692b3bcd] JLLWrappers v1.4.1
  [682c06a0] JSON v0.21.3
  [0f8b85d8] JSON3 v1.12.0
  [aa1ae85d] JuliaInterpreter v0.9.23
  [70703baa] JuliaSyntax v0.3.4
  [1d6d02ad] LeftChildRightSiblingTrees v0.2.0
  [e6f89c97] LoggingExtras v1.0.0
  [6f1432cf] LoweredCodeUtils v2.3.0
  [6ac74813] MD5 v0.2.1
  [1914dd2f] MacroTools v0.5.10
  [739be429] MbedTLS v1.1.7
  [c03570c3] Memoize v0.4.4
  [78c3b35d] Mocking v0.7.6
  [4d8831e6] OpenSSL v1.3.3
  [bac558e1] OrderedCollections v1.6.0
  [e4faabce] PProf v2.2.2
  [69de0a69] Parsers v2.5.8
  [2dfb63ee] PooledArrays v1.4.2
  [21216c6a] Preferences v1.3.0
  [92933f4c] ProgressMeter v1.7.2
  [3349acd9] ProtoBuf v1.0.9
  [9c30249a] RAI v0.2.4 `~/.julia/dev/RAI`
  [3cdcf5f2] RecipesBase v1.3.3
  [189a3867] Reexport v1.2.2
  [ae029012] Requires v1.3.0
  [295af30f] Revise v3.5.2
  [6c6a2e73] Scratch v1.2.0
  [91c51154] SentinelArrays v1.3.18
  [777ac1f9] SimpleBufferStream v1.1.0
  [66db9d55] SnoopPrecompile v1.0.3
  [856f2bd8] StructTypes v1.10.0
  [3783bdb8] TableTraits v1.0.1
  [bd369af6] Tables v1.10.1
  [b718987f] TextWrap v1.0.1
  [f269a46b] TimeZones v1.9.1
  [3bb67fe8] TranscodingStreams v0.9.11
  [d265eb64] TypedSyntax v1.1.7
  [5c2747f8] URIs v1.4.2
  [b8c1c048] WidthLimitedIO v1.0.1
  [76eceee3] WorkerUtilities v1.6.1
  [228000da] XMLDict v0.4.1
  [6e34b625] Bzip2_jll v1.0.8+0
  [83423d85] Cairo_jll v1.16.1+1
  [2e619515] Expat_jll v2.4.8+0
  [a3f928ae] Fontconfig_jll v2.13.93+0
  [d7e528f0] FreeType2_jll v2.10.4+0
  [559328eb] FriBidi_jll v1.0.10+0
  [78b55507] Gettext_jll v0.21.0+0
  [7746bdde] Glib_jll v2.74.0+2
  [3b182d85] Graphite2_jll v1.3.14+0
  [3c863552] Graphviz_jll v2.50.0+1
  [2e76f6c2] HarfBuzz_jll v2.8.1+1
  [dd4b983a] LZO_jll v2.10.1+0
⌅ [e9f186c6] Libffi_jll v3.2.2+1
  [d4300ac3] Libgcrypt_jll v1.8.7+0
  [7add5ba3] Libgpg_error_jll v1.42.0+0
  [94ce4f54] Libiconv_jll v1.16.1+2
  [4b2f31a3] Libmount_jll v2.35.0+0
  [38a345b3] Libuuid_jll v2.36.0+0
  [5ced341a] Lz4_jll v1.9.3+0
  [a4b94fbf] NodeJS_16_jll v16.15.0+0
  [458c3c95] OpenSSL_jll v1.1.20+0
  [36c8627f] Pango_jll v1.50.9+0
  [30392449] Pixman_jll v0.40.1+0
  [02c8fc9c] XML2_jll v2.10.3+0
  [aed1982a] XSLT_jll v1.1.34+0
  [4f6342f7] Xorg_libX11_jll v1.6.9+4
  [0c0b7dd1] Xorg_libXau_jll v1.0.9+4
  [a3789734] Xorg_libXdmcp_jll v1.1.3+4
  [1082639a] Xorg_libXext_jll v1.3.4+4
  [ea2f1a96] Xorg_libXrender_jll v0.9.10+4
  [14d82f49] Xorg_libpthread_stubs_jll v0.1.0+3
  [c7cfdc94] Xorg_libxcb_jll v1.13.0+3
  [c5fb5394] Xorg_xtrans_jll v1.4.0+3
  [3161d3a3] Zstd_jll v1.5.4+0
  [d59cea07] azurite_jll v3.18.0+0
  [b53b4c65] libpng_jll v1.6.38+0
  [ad83bbfc] minio_jll v1.0.0+1
  [cf2c5f97] pprof_jll v1.0.1+0
  [0dad84c5] ArgTools v1.1.1
  [56f22d72] Artifacts
  [2a0f44e3] Base64
  [ade2ca70] Dates
  [8ba89e20] Distributed
  [f43a241f] Downloads v1.6.0
  [7b1f6079] FileWatching
  [9fa8497b] Future
  [b77e0a4c] InteractiveUtils
  [4af54fe1] LazyArtifacts
  [b27032c2] LibCURL v0.6.3
  [76f85450] LibGit2
  [8f399da3] Libdl
  [37e2e46d] LinearAlgebra
  [56ddb016] Logging
  [d6f4376e] Markdown
  [a63ad114] Mmap
  [ca575930] NetworkOptions v1.2.0
  [44cfe95a] Pkg v1.8.0
  [de0858da] Printf
  [9abbd945] Profile
  [3fa0cd96] REPL
  [9a3f8284] Random
  [ea8e919c] SHA v0.7.0
  [9e88b42a] Serialization
  [6462fe0b] Sockets
  [2f01184e] SparseArrays
  [10745b16] Statistics
  [fa267f1f] TOML v1.0.0
  [a4e569a6] Tar v1.10.1
  [8dfed614] Test
  [cf7118a7] UUIDs
  [4ec0a83e] Unicode
  [e66e0078] CompilerSupportLibraries_jll v0.5.2+0
  [deac9b47] LibCURL_jll v7.84.0+0
  [29816b5a] LibSSH2_jll v1.10.2+0
  [c8ffd9c3] MbedTLS_jll v2.28.0+0
  [14a3606d] MozillaCACerts_jll v2022.2.1
  [4536629a] OpenBLAS_jll v0.3.20+0
  [efcefdf7] PCRE2_jll v10.40.0+0
  [83775a58] Zlib_jll v1.2.12+3
  [8e850b90] libblastrampoline_jll v5.1.1+0
  [8e850ede] nghttp2_jll v1.48.0+0
  [3f19e933] p7zip_jll v17.4.0+0
Info Packages marked with ⌅ have new versions available but compatibility constraints restrict them from upgrading. To see why use `status --outdated -m`

(@v1.8) pkg>
```

Thank you for this _amazing_ package! The latest improvements with showing the source code are absolutely stunning. 🤩

--- a/TypedSyntax/Project.toml
+++ b/TypedSyntax/Project.toml
@@ -13,6 +13,7 @@ JuliaSyntax = "0.3.2"
 julia = "1"
 
 [extras]
+Dates = "ade2ca70-3891-5945-98fb-dc099432e06a"
 InteractiveUtils = "b77e0a4c-d291-57a0-90e8-8db25a27a240"
 Logging = "56ddb016-857b-54e1-b83d-db4d58db5568"
 MethodAnalysis = "85b6ec6f-f7df-4429-9514-a64bcd9ee824"
@@ -20,4 +21,4 @@ ProgressMeter = "92933f4c-e287-5a05-a399-4b506db050ca"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 
 [targets]
-test = ["InteractiveUtils", "Logging", "MethodAnalysis", "ProgressMeter", "Test"]
+test = ["Dates", "InteractiveUtils", "Logging", "MethodAnalysis", "ProgressMeter", "Test"]

--- a/TypedSyntax/src/node.jl
+++ b/TypedSyntax/src/node.jl
@@ -333,7 +333,7 @@ end
 function num_positional_args(tsn::AbstractSyntaxNode)
     TypedSyntax.is_function_def(tsn) || return 0
     sig, _ = children(tsn)
-    if kind(sig) == K"where"
+    while kind(sig) == K"where"
         sig = child(sig, 1)
     end
     @assert kind(sig) == K"call"

--- a/TypedSyntax/src/show.jl
+++ b/TypedSyntax/src/show.jl
@@ -52,7 +52,7 @@ function Base.printstyled(io::IO, rootnode::MaybeTypedSyntaxNode;
 end
 Base.printstyled(rootnode::MaybeTypedSyntaxNode; kwargs...) = printstyled(stdout, rootnode; kwargs...)
 
-ndigits_linenumbers(node, idxend = last_byte(node)) = ndigits(node.source.first_line + nlines(node.source, idxend))
+ndigits_linenumbers(node, idxend = last_byte(node)) = ndigits(node.source.first_line + nlines(node.source, idxend) - 1)
 
 function show_src_expr(io::IO, node::MaybeTypedSyntaxNode, position::Int, pre::String, pre2::String; type_annotations::Bool=true, iswarn::Bool=false, hide_type_stable::Bool=false, nd::Int)
     _lastidx = last_byte(node)

--- a/TypedSyntax/test/runtests.jl
+++ b/TypedSyntax/test/runtests.jl
@@ -1,6 +1,6 @@
 using JuliaSyntax: JuliaSyntax, SyntaxNode, children, child, sourcetext, kind, @K_str
 using TypedSyntax: TypedSyntax, TypedSyntaxNode, getsrc
-using InteractiveUtils, Test
+using Dates, InteractiveUtils, Test
 
 has_name_typ(node, name::Symbol, @nospecialize(T)) = kind(node) == K"Identifier" && node.val === name && node.typ === T
 has_name_notyp(node, name::Symbol) = has_name_typ(node, name, nothing)
@@ -532,6 +532,9 @@ include("test_module.jl")
     # issue #397
     tsn = TypedSyntaxNode(TSN.f397, (typeof(view([1,2,3], 1:2)),))
     @test TypedSyntax.num_positional_args(tsn) == 2 # the function is arg1, x is arg2
+    # issue #426
+    tsn = TypedSyntaxNode(getindex, (typeof(TSN.T426), Type{Year},))
+    @test TypedSyntax.num_positional_args(tsn) == 3
 
     # Display
     tsn = TypedSyntaxNode(TSN.mysin, (Int,))
@@ -589,10 +592,10 @@ include("test_module.jl")
         printstyled(io, obj; hide_type_stable=false)
     end
     @test str === """
-        4 function simplef(a::Int64, b::Float64)::Float64
-        5     z::Int64 = (a::Int64 * a::Int64)::Int64
-        6     return (z::Int64 + b::Float64)::Float64
-        7 end"""
+        5 function simplef(a::Int64, b::Float64)::Float64
+        6     z::Int64 = (a::Int64 * a::Int64)::Int64
+        7     return (z::Int64 + b::Float64)::Float64
+        8 end"""
     tsn = TypedSyntaxNode(TSN.myabs, (Float64,))
     str = sprint(tsn; context=:color=>false) do io, obj
         printstyled(io, obj; hide_type_stable=false)

--- a/TypedSyntax/test/test_module.jl
+++ b/TypedSyntax/test/test_module.jl
@@ -1,4 +1,5 @@
 module TSN
+using Dates
 
 # with two uses of the same slot in the same call. Must start on line 4 (or update the corresponding test)
 function simplef(a, b)
@@ -205,5 +206,11 @@ end
 struct ReadOnly end
 fname = :eltype
 @eval Base.@propagate_inbounds @inline Base.$fname(::ReadOnly) = Int
+
+# Issue #426
+const T426 = Dict{Type{<:Dates.Period}, Bool}(
+    Dates.Year => true,
+    Dates.Month => false,
+)
 
 end

